File: php/src/SlhDsa/Fors.php
<?php

declare(strict_types=1);

namespace PQC\SlhDsa;

/**
 * FORS (Forest of Random Subsets) for SLH-DSA.
 * FIPS 205 Section 8.
 */
final class Fors
{
    /**
     * Generate a FORS private key value.
     */
    private static function skGen(string $skSeed, string $pkSeed, Address $adrs, int $idx, int $n): string
    {
        $skAdrs = $adrs->copy();
        $skAdrs->setType(Address::FORS_PRF);
        $skAdrs->setKeyPairAddress($adrs->getKeyPairAddress());
        $skAdrs->setTreeIndex($idx);
        $skAdrs->setTreeHeight(0);

        return SlhHash::prf($pkSeed, $skSeed, $skAdrs, $n);
    }

    /**
     * Compute a node in a FORS tree.
     */
    private static function treeNode(
        string $skSeed,
        string $pkSeed,
        int $targetNode,
        int $targetHeight,
        Address $adrs,
        int $treeIdx,
        int $n
    ): string {
        if ($targetHeight === 0) {
            $sk = self::skGen($skSeed, $pkSeed, $adrs, $treeIdx * (1 << 0) + $targetNode, $n);
            $leafAdrs = $adrs->copy();
            $leafAdrs->setType(Address::FORS_TREE);
            $leafAdrs->setKeyPairAddress($adrs->getKeyPairAddress());
            $leafAdrs->setTreeHeight(0);
            $leafAdrs->setTreeIndex($treeIdx * 1 + $targetNode);

            return SlhHash::F($pkSeed, $leafAdrs, $sk, $n);
        }

        $lNode = self::treeNode($skSeed, $pkSeed, 2 * $targetNode, $targetHeight - 1, $adrs, $treeIdx, $n);
        $rNode = self::treeNode($skSeed, $pkSeed, 2 * $targetNode + 1, $targetHeight - 1, $adrs, $treeIdx, $n);

        $nodeAdrs = $adrs->copy();
        $nodeAdrs->setType(Address::FORS_TREE);
        $nodeAdrs->setKeyPairAddress($adrs->getKeyPairAddress());
        $nodeAdrs->setTreeHeight($targetHeight);
        $nodeAdrs->setTreeIndex($treeIdx * (1 << $targetHeight) + $targetNode);

        return SlhHash::H($pkSeed, $nodeAdrs, $lNode . $rNode, $n);
    }

    /**
     * Generate FORS signature.
     * FIPS 205 Algorithm 14.
     *
     * @param string $md Message digest
     * @param string $skSeed Secret seed
     * @param string $pkSeed Public seed
     * @param Address $adrs Address
     * @param array $params Parameters
     * @return string FORS signature
     */
    public static function sign(
        string $md,
        string $skSeed,
        string $pkSeed,
        Address $adrs,
        array $params
    ): string {
        $n = $params['n'];
        $k = $params['k'];
        $a = $params['a'];

        // Split md into k indices of a bits each
        $indices = self::messageToIndices($md, $k, $a);

        $sig = '';
        for ($i = 0; $i < $k; $i++) {
            $idx = $indices[$i];

            // Secret value
            $sig .= self::skGen($skSeed, $pkSeed, $adrs, $i * (1 << $a) + $idx, $n);

            // Authentication path
            for ($j = 0; $j < $a; $j++) {
                $s = ($idx >> $j) ^ 1;
                $sig .= self::treeNode($skSeed, $pkSeed, $s, $j, $adrs, $i, $n);
            }
        }

        return $sig;
    }

    /**
     * Compute FORS public key from signature.
     * FIPS 205 Algorithm 15.
     */
    public static function pkFromSig(
        string $sig,
        string $md,
        string $pkSeed,
        Address $adrs,
        array $params
    ): string {
        $n = $params['n'];
        $k = $params['k'];
        $a = $params['a'];

        $indices = self::messageToIndices($md, $k, $a);

        $roots = '';
        $offset = 0;

        for ($i = 0; $i < $k; $i++) {
            $idx = $indices[$i];

            // Get sk value from sig
            $sk = substr($sig, $offset, $n);
            $offset += $n;

            // Compute leaf
            $leafAdrs = $adrs->copy();
            $leafAdrs->setType(Address::FORS_TREE);
            $leafAdrs->setKeyPairAddress($adrs->getKeyPairAddress());
            $leafAdrs->setTreeHeight(0);
            $leafAdrs->setTreeIndex($i * (1 << $a) + $idx);

            $node = SlhHash::F($pkSeed, $leafAdrs, $sk, $n);

            // Climb the tree
            for ($j = 0; $j < $a; $j++) {
                $authJ = substr($sig, $offset, $n);
                $offset += $n;

                $nodeAdrs = $adrs->copy();
                $nodeAdrs->setType(Address::FORS_TREE);
                $nodeAdrs->setKeyPairAddress($adrs->getKeyPairAddress());
                $nodeAdrs->setTreeHeight($j + 1);

                if ((($idx >> $j) & 1) === 0) {
                    $nodeAdrs->setTreeIndex($i * (1 << ($j + 1)) + ($idx >> ($j + 1)));
                    $node = SlhHash::H($pkSeed, $nodeAdrs, $node . $authJ, $n);
                } else {
                    $nodeAdrs->setTreeIndex($i * (1 << ($j + 1)) + ($idx >> ($j + 1)));
                    $node = SlhHash::H($pkSeed, $nodeAdrs, $authJ . $node, $n);
                }
            }

            $roots .= $node;
        }

        // Compress FORS roots into single public key
        $forsRootAdrs = $adrs->copy();
        $forsRootAdrs->setType(Address::FORS_ROOTS);
        $forsRootAdrs->setKeyPairAddress($adrs->getKeyPairAddress());

        return SlhHash::Tl($pkSeed, $forsRootAdrs, $roots, $n);
    }

    /**
     * Split message digest into k indices of a bits each.
     */
    private static function messageToIndices(string $md, int $k, int $a): array
    {
        $indices = [];
        $bits = 0;
        $buffer = 0;
        $pos = 0;

        for ($i = 0; $i < $k; $i++) {
            while ($bits < $a) {
                if ($pos < strlen($md)) {
                    $buffer = ($buffer << 8) | ord($md[$pos++]);
                    $bits += 8;
                } else {
                    $buffer <<= 8;
                    $bits += 8;
                }
            }
            $bits -= $a;
            $indices[] = ($buffer >> $bits) & ((1 << $a) - 1);
        }

        return $indices;
    }
}

File: java/src/main/java/com/pqc/slhdsa/Fors.java
package com.pqc.slhdsa;

/**
 * FORS (Forest of Random Subsets) per FIPS 205 Section 8.
 */
public final class Fors {

    private Fors() {}

    /**
     * fors_SKgen(SK.seed, PK.seed, ADRS, idx): Generate a FORS private key value
     * (Algorithm 11 in FIPS 205).
     */
    public static byte[] skGen(byte[] skSeed, byte[] pkSeed, Address adrs, int idx,
                                SlhHash hash, int n) {
        Address skAdrs = new Address(adrs);
        skAdrs.setType(Address.FORS_PRF);
        skAdrs.setKeyPairAddress(adrs.getKeyPairAddress());
        skAdrs.setTreeIndex(idx);
        return hash.prf(pkSeed, skSeed, skAdrs, n);
    }

    /**
     * fors_node(SK.seed, i, z, PK.seed, ADRS): Compute the root of a
     * Merkle subtree in FORS (Algorithm 12 in FIPS 205).
     */
    public static byte[] node(byte[] skSeed, int i, int z, byte[] pkSeed, Address adrs,
                               SlhParams params, SlhHash hash) {
        int n = params.n;

        if (z == 0) {
            byte[] sk = skGen(skSeed, pkSeed, adrs, i, hash, n);
            Address leafAdrs = new Address(adrs);
            leafAdrs.setType(Address.FORS_TREE);
            leafAdrs.setKeyPairAddress(adrs.getKeyPairAddress());
            leafAdrs.setTreeHeight(0);
            leafAdrs.setTreeIndex(i);
            return hash.f(pkSeed, leafAdrs, sk, n);
        }

        byte[] lNode = node(skSeed, 2 * i, z - 1, pkSeed, adrs, params, hash);
        byte[] rNode = node(skSeed, 2 * i + 1, z - 1, pkSeed, adrs, params, hash);

        Address treeAdrs = new Address(adrs);
        treeAdrs.setType(Address.FORS_TREE);
        treeAdrs.setKeyPairAddress(adrs.getKeyPairAddress());
        treeAdrs.setTreeHeight(z);
        treeAdrs.setTreeIndex(i);

        return hash.h(pkSeed, treeAdrs, SlhUtils.concat(lNode, rNode), n);
    }

    /**
     * fors_sign(md, SK.seed, PK.seed, ADRS): Generate a FORS signature
     * (Algorithm 13 in FIPS 205).
     */
    public static byte[] sign(byte[] md, byte[] skSeed, byte[] pkSeed, Address adrs,
                               SlhParams params, SlhHash hash) {
        int n = params.n;
        int k = params.k;
        int a = params.a;

        // Split md into k a-bit unsigned integers
        int[] indices = SlhUtils.base2b(md, a, k);

        byte[] sigFors = new byte[k * (a + 1) * n];
        int offset = 0;

        for (int i = 0; i < k; i++) {
            int idx = indices[i];

            // Private key value
            byte[] sk = skGen(skSeed, pkSeed, adrs, i * (1 << a) + idx, hash, n);
            System.arraycopy(sk, 0, sigFors, offset, n);
            offset += n;

            // Authentication path
            for (int j = 0; j < a; j++) {
                int s = (idx >>> j) ^ 1;
                byte[] authNode = node(skSeed, i * (1 << (a - j)) + s, j, pkSeed, adrs, params, hash);
                System.arraycopy(authNode, 0, sigFors, offset, n);
                offset += n;
            }
        }
        return sigFors;
    }

    /**
     * fors_pkFromSig(sig_fors, md, PK.seed, ADRS): Derive the FORS public key
     * from a FORS signature (Algorithm 14 in FIPS 205).
     */
    public static byte[] pkFromSig(byte[] sigFors, byte[] md, byte[] pkSeed, Address adrs,
                                    SlhParams params, SlhHash hash) {
        int n = params.n;
        int k = params.k;
        int a = params.a;

        int[] indices = SlhUtils.base2b(md, a, k);

        byte[] roots = new byte[k * n];
        int sigOffset = 0;

        for (int i = 0; i < k; i++) {
            int idx = indices[i];

            // Retrieve sk value from signature
            byte[] sk = SlhUtils.slice(sigFors, sigOffset, n);
            sigOffset += n;

            // Compute leaf node
            Address leafAdrs = new Address(adrs);
            leafAdrs.setType(Address.FORS_TREE);
            leafAdrs.setKeyPairAddress(adrs.getKeyPairAddress());
            leafAdrs.setTreeHeight(0);
            leafAdrs.setTreeIndex(i * (1 << a) + idx);
            byte[] node0 = hash.f(pkSeed, leafAdrs, sk, n);

            // Walk authentication path
            for (int j = 0; j < a; j++) {
                byte[] authJ = SlhUtils.slice(sigFors, sigOffset, n);
                sigOffset += n;

                Address treeAdrs = new Address(adrs);
                treeAdrs.setType(Address.FORS_TREE);
                treeAdrs.setKeyPairAddress(adrs.getKeyPairAddress());
                treeAdrs.setTreeHeight(j + 1);

                if (((idx >>> j) & 1) == 0) {
                    treeAdrs.setTreeIndex(i * (1 << (a - j - 1)) + (idx >>> (j + 1)));
                    node0 = hash.h(pkSeed, treeAdrs, SlhUtils.concat(node0, authJ), n);
                } else {
                    treeAdrs.setTreeIndex(i * (1 << (a - j - 1)) + (idx >>> (j + 1)));
                    node0 = hash.h(pkSeed, treeAdrs, SlhUtils.concat(authJ, node0), n);
                }
            }
            System.arraycopy(node0, 0, roots, i * n, n);
        }

        // Compute FORS public key as T_k(PK.seed, ADRS, root_0 || ... || root_{k-1})
        Address forsPkAdrs = new Address(adrs);
        forsPkAdrs.setType(Address.FORS_ROOTS);
        forsPkAdrs.setKeyPairAddress(adrs.getKeyPairAddress());

        return hash.tl(pkSeed, forsPkAdrs, roots, n);
    }
}
